packages/nitro-protocol/test/contracts/MultiAssetHolder/claim.test.ts

```diff
@@ -40,24 +41,28 @@ const reason6 = 'targetAsset != guaranteeAsset';
+    ${' 9. (all) straight-through guarantee, 3 destinations'}         | ${{g: 5}}  | ${['I', 'A', 'B']}    | ${{I: 5, A: 5, B: 5}} | ${[]}   | ${{I: 0, A: 5, B: 5}} | ${{g: 0}}        | ${{I: 5}}       | ${undefined}
+    ${'10. (all) swap guarantee,             2 destinations'}         | ${{g: 5}}  | ${['B', 'A']}         | ${{A: 5, B: 5}}       | ${[]}   | ${{A: 5, B: 0}}       | ${{g: 0}}        | ${{B: 5}}       | ${undefined}
+    ${'11. (all) swap guarantee,             3 destinations'}         | ${{g: 5}}  | ${['I', 'B', 'A']}    | ${{I: 5, A: 5, B: 5}} | ${[]}   | ${{I: 0, A: 5, B: 5}} | ${{g: 0}}        | ${{I: 5}}       | ${undefined}
+    ${'12. (all) straight-through guarantee, 2 destinations'}         | ${{g: 5}}  | ${['A', 'B']}         | ${{A: 5, B: 5}}       | ${[]}   | ${{A: 0, B: 5}}       | ${{g: 0}}        | ${{A: 5}}       | ${undefined}
+    ${'13. (all) allocation not on chain'}                            | ${{g: 5}}  | ${['B', 'A']}         | ${{}}                 | ${[]}   | ${{}}                 | ${{g: 0}}        | ${{B: 5}}       | ${reason5}
+    ${'14. (all) guarantee not on chain'}                             | ${{g: 5}}  | ${[]}                 | ${{A: 5, B: 5}}       | ${[]}   | ${{A: 5, B: 5}}       | ${{g: 0}}        | ${{B: 5}}       | ${reason5}
+    ${'15. (all) swap guarantee, overfunded, 2 destinations'}         | ${{g: 12}} | ${['B', 'A']}         | ${{A: 5, B: 5}}       | ${[]}   | ${{A: 0, B: 0}}       | ${{g: 2}}        | ${{A: 5, B: 5}} | ${undefined}
+    ${'16. (all) underspecified guarantee, overfunded      '}         | ${{g: 12}} | ${['B']}              | ${{A: 5, B: 5}}       | ${[]}   | ${{A: 5, B: 0}}       | ${{g: 7}}        | ${{A: 5, B: 5}} | ${undefined}
+    ${'17. guarantee and target assets do not match'}                 | ${{g: 1}}  | ${['A', 'B']}         | ${{A: 5, B: 5}}       | ${[1]}  | ${{A: 4, B: 5}}       | ${{g: 0}}        | ${{A: 1}}       | ${reason6}
+    ${'18. absent guarantee entry, indices=[], bad guarantee order'}  | ${{g: 10}} | ${['A', 'I', 'x']}    | ${{x: 10, I: 10}}     | ${[]}   | ${{x: 10, I: 0}}      | ${{g: 0}}        | ${{I: 10}}      | ${undefined}
+    ${'19. absent guarantee entry, indices=[1], bad guarantee order'} | ${{g: 10}} | ${['A', 'I', 'x']}    | ${{x: 10, I: 10}}     | ${[1]}  | ${{x: 10, I: 0}}      | ${{g: 0}}        | ${{I: 10}}      | ${undefined}
+    ${'20. absent guarantee entry, indices=[]'}                       | ${{g: 10}} | ${['A', 'x', 'I']}    | ${{x: 10, I: 10}}     | ${[]}   | ${{x: 0, I: 10}}      | ${{g: 0, x: 10}} | ${{}}           | ${undefined}
+    ${'21. absent guarantee entry, indices=[1]'}                      | ${{g: 10}} | ${['A', 'x', 'I']}    | ${{x: 10, I: 10}}     | ${[0]}  | ${{x: 0, I: 10}}      | ${{g: 0, x: 10}} | ${{}}           | ${undefined}
```

Comment:
Am I right to understand that this test case would have failed before fixing the bug?

Reply:
yes

Reply:
I checked that [here](https://github.com/statechannels/statechannels/pull/3745), and found that the test cases that were added did not fail on master. Are you sure there's a bug in the current implementation?

After reviewing the implementation of `computeNewAllocationFromGuarantee` as well as the current implementation of `_claim`, I am not convinced that there's a bug:
- `payouts[k]` is set [here](https://github.com/statechannels/statechannels/blob/f0cc53fdc3ecf64c12535f9f83e86bf7a0f37f0d/packages/nitro-protocol/contracts/MultiAssetHolder.sol#L325). Note that `payouts[k]` is marking the amount that should get paid to `allocation[indices[k]].destination`
- [later](https://github.com/statechannels/statechannels/blob/f0cc53fdc3ecf64c12535f9f83e86bf7a0f37f0d/packages/nitro-protocol/contracts/MultiAssetHolder.sol#L371), when deciding where to send `payouts[j]`, we set `destination = allocation[indices[j]]`

This is more complicated than it needs to be, but it seems to be correct?

Reply:
Those tests should fail, but don't fail. I'll work on fixing that.

Reply:
Why should the tests fail? Is my understanding outlined above incorrect?

Reply:
Let's tale one test case in particular [(18.)](https://github.com/statechannels/statechannels/pull/3745/files?diff=unified&w=1).

We are claiming a guarantee of `[A,I,x]` (funded with 10) targeting an outcome of `{x:10,I:10}` with `indices=[]` signalling our wish to claim all of the target allocations. `I` should be paid 10 and `x` should not be paid anything. Instead, `x` is paid 10 and `I` is not paid anything. 

The problem with the implementation is that `k` remains at its initial value of `0`; so the payout calculated for `I` is stored in the slot for `x`. The outer function then pays the coins to `x`.

Unfortunately the tests are not actually checking the payouts, hence the tests pass. I have remedied that on #3747.